Extended Parmest Capability for weighted SSE objective

slilonfe5 opened the pull request:
<!-- ##################################################################### -->
<!-- PLEASE READ BEFORE OPENING THIS PULL REQUEST -->

<!-- All changes must adhere to PEP8 standards as enforced by Black. -->
<!-- If your changes do NOT adhere, the test suite will fail. -->
<!-- Please read our Contributing guide for instructions on how to apply these standards. -->
<!-- Contributing Guide: https://pyomo.readthedocs.io/en/stable/contribution_guide.html -->
<!-- ##################################################################### -->

<!-- DO NOT DELETE OR IGNORE THIS TEMPLATE. Failing to adhere to this template and provide the necessary information may lead to your Pull Request being closed without consideration. -->

## Fixes # .

## Summary/Motivation:
Currently, the Parmest SSE objective does not support measurements in different units. This work adds new capabilities to Parmest, including weighted SSE to handle measurements in different units, and more robust covariance matrix calculation methods for more accurate uncertainty quantification. This work also enables the calculation of the covariance matrix using a user-supplied measurement error standard deviation.

## Changes proposed in this PR:
- Added a weighted SSE objective
- Added two covariance matrix calculation methods for both the SSE and weighted SSE objectives
- Enabled calculation of the covariance matrix using a user-supplied measurement error standard deviation
- Added a separate function for the covariance matrix estimation

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

slilonfe5 commented:
@adowling2 @djlaky

adowling2 commented:
@slilonfe5 Here is some quick feedback

`compute_jacobian` function
- [ ] Make this a private method by adding `_` to the function name
- [ ] Add as an argument to the function `relative_perturbation`
- [ ] In the document string, explain this is using forward (?) finite difference
- [ ] Add as an argument the solver object. You can make the default Ipopt.

Feedback on the `compute_FIM` method:
- [ ] Add `relative_tolerance` and `solver` as arguments
- [ ] Also add a check that `error_list` must be the same length as `y_hat_list`
- [ ] Add a debugging step for the linear algebra error, compute the condition number of the Jacobian matrix and print it out
- [ ] Why would you ever get a linear algebra error for just matrix multiplication? Is this check even needed?

slilonfe5 commented:
@adowling2 @djlaky

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1726,10 +1891,65 @@ def _Q_opt(
-                cov = pd.DataFrame(
-                    cov, index=thetavals.keys(), columns=thetavals.keys()
-                )
+                if self.obj_function == 'SSE': # covariance calculation for measurements in the same unit
```
Are you adding the same code in two places? If so, can we streamline this into a function to avoid duplicate code?

adowling2 replied:
Suggestion: remove this. Do not add this capability for the depreciated interface.

adowling2 replied:
You can even add an error message that computing the covariance is no longer supported in the deprecated interface.

adowling2 commented:
@slilonfe5 Once you have the tests ready, tag us for feedback. Also, I think you can skip adding this to the depreciated class.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -229,11 +229,119 @@ def _experiment_instance_creation_callback(
+# Calculate the sensitivity of measured variables to parameters using central finite difference
+def _compute_jacobian(model, relative_perturbation, solver_option="ipopt"):
+    """
+    Computes the Jacobian matrix using central finite difference scheme
```
Jacobian matrix of measurement predictions with respect to changes in parameter values.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -229,11 +229,119 @@ def _experiment_instance_creation_callback(
+    return J
+
+# compute the Fisher information matrix of the estimated parameters
+def compute_FIM(model, relative_perturbation, solver_option="ipopt"):
```
Allow the user to provide a solver object instead of the solver name. Likewise, allow the user to select from several methods including: (1) central finite difference, (2) kaug, and (3) reduced Hessian.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -578,10 +688,60 @@ def _Q_opt(
-                cov = pd.DataFrame(
-                    cov, index=thetavals.keys(), columns=thetavals.keys()
-                )
+                if self.obj_function == 'SSE':
```
Suggestion: move this to a separate function. Also, aren't compute_FIM and compute_cov redundant, as they are inverses of each other?

adowling2 reviewed (commented):
Here is more feedback as you work on getting this ready for the Pyomo team to review.

slilonfe5 commented:
@adowling2 @djlaky I have created a separate method (`cov_est`) for computing the covariance matrix, supporting three calculation methods (`jacobian`, `kaug`, and `reduced_hessian`). I implemented these covariance calculation methods for both the `SSE` and `SSE_weighted` objectives. Lastly, as you suggested, I did not add the new capability to the deprecated interface. 

I tested these with three examples (2 steady state and 1 dynamic), and all work well. I'm yet to write the test file for these.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+def SSE_weighted(model):
+    """
+    Weighted sum of squared error between the model prediction of measured variables and data values,
+    assuming Gaussian i.i.d errors, with measurement error standard deviation defined in the annotated Pyomo model
```
Picky, but make sure you use `i.i.d.` with a period at the end.

Also, the description is usually what the function does. I would probably say something about returning an expression that is used to compute the SSE or weighted SSE.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+    assuming Gaussian i.i.d errors, with measurement error standard deviation defined in the annotated Pyomo model
+    """
+    if not hasattr(model, "measurement_error"):
+        raise AttributeError('The model must have a `measurement_error` attribute. Please define it')
```
Check to see what the errors are in pyomo.DoE. It may be good to standardize this error between both modules.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+    if not hasattr(model, "measurement_error"):
+        raise AttributeError('The model must have a `measurement_error` attribute. Please define it')
+
+    if all(model.measurement_error[y_hat] is None for y_hat in model.experiment_outputs):
```
Does this throw a key error if `y_hat` is not in `measurement_error`?

You probably need to check 2 things: (1) all `y_hat` are in `measurement_error`, and (2) `measurement_error` has values for each `y_hat`.

You already check (2), but (1) seems to be unchecked and may cause you some issues.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+            for y_hat, y in model.experiment_outputs.items()
+        )
+        return expr
+    else:
```
It may be better to lump this in with the first case. i.e., just check if `all(model.measurement_error[y_hat] is not None for y_hat in model.experiment_outputs)`, then if this isn't true, say that "one or more values are missing from `measurement_error`.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+# Compute the Jacobian matrix of measured variables with respect to the parameters
+def _compute_jacobian(experiment, thetavals, absolute_perturbation, solver, tee):
+    """
+    Computes the Jacobian matrix of the measured variables with respect to the parameters
```
Probably need to specify "unknown parameters"

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+    using central finite difference scheme
+
+    Arguments:
+        experiment: Estimator class object that contains the model for a particular experimental condition
```
You should include the expected type of input for each of these inputs.

Then you should have a soft test to make sure they are what you think they are. It is often best to include a default value for some of these like `None` and check if they have been supplied or not; and if they have been supplied, if they are the correct type.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+    # resolve the model with the estimated parameters
+    solver = pyo.SolverFactory(solver)
+    solver.solve(model, tee=tee)
+
```
Should check that the model exits with a solved status.

adowling2 replied:
Done.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+    model = experiment.get_labeled_model().clone()
+
+    # fix the parameter values to the optimal values estimated
+    params = [k for k, v in model.unknown_parameters.items()]
```
Is it assumed that unknown parameters exists on the model? You may need to check for each of these keywords to make sure they are defined.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+    solver.solve(model, tee=tee)
+
+    # get the measured variables
+    y_hat_list = [y_hat for y_hat, y in model.experiment_outputs.items()]
```
Same thing with checking `experiment_outputs` exists.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+    y_hat_list = [y_hat for y_hat, y in model.experiment_outputs.items()]
+
+    # get the parameters
+    params = [k for k, v in model.unknown_parameters.items()]
```
Why do you get the params again? We already had then earlier.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+        param.fix(orig_value + absolute_perturbation)
+
+        # solve model
+        solver.solve(model, tee=tee)
```
Make sure this exits as solved (as before).

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+        param.fix(orig_value - absolute_perturbation)
+
+        # resolve model
+        solver.solve(model, tee=tee)
```
Make sure this exits as solved (as before).

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+        orig_value = param_values[i]
+
+        # Forward perturbation
+        param.fix(orig_value + absolute_perturbation)
```
I would probably use a relative perturbation here in the case that parameters have different scales.

You just need to update your input description and use it to multiply the original value a couple lines above.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+    Returns:
+        cov: covariance matrix of the estimated parameters
+    """
+    if method == "jacobian":
```
They may ask you to use an object like the one in DoE for objective types: https://github.com/Pyomo/pyomo/blob/86f5f05ea3796be4e79c06772c71f330c19041c4/pyomo/contrib/doe/doe.py#L65

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+        FIM = np.sum(FIM_all_exp, axis=0)
+
+        # covariance matrix
+        cov = np.linalg.inv(FIM)
```
Same as previous comment.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+    measurement errors standard deviation defined in the annotated Pyomo model
+
+    Arguments:
+        experiment: Estimator class object that contains the model
```
Same comment about including type expectations for these inputs.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -226,14 +228,310 @@ def _experiment_instance_creation_callback(
+        if len(error_list) != len(y_hat_list):
+            raise ValueError("Experiment outputs and measurement errors are not the same length")
+
+        # calculate the FIM
```
May be helpful to include a reference, even if it references a future manuscript

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -428,6 +727,8 @@ def _create_parmest_model(self, experiment_number):
             if self.obj_function == 'SSE':
                 second_stage_rule = SSE
+            elif self.obj_function == 'SSE_weighted':
+                second_stage_rule = SSE_weighted
```
Again, may want to make an objective list like that we have in `pyomo.doe`:

https://github.com/Pyomo/pyomo/blob/86f5f05ea3796be4e79c06772c71f330c19041c4/pyomo/contrib/doe/doe.py#L65-L68

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -508,36 +807,28 @@ def _Q_opt(
 
         # Solve the extensive form with ipopt
         if solver == "ef_ipopt":
-            if not calc_cov:
```
Make sure it is clear that this is being moved. This could be a big change for people that already use the tool.....

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,19 +880,125 @@ def _Q_opt(
+            # check if user defined the measurement_error attribute
+            if hasattr(model, "measurement_error"):
+                # get the measurement errors
+                meas_error = [model.measurement_error[y_hat] for y_hat, y in model.experiment_outputs.items()]
```
This probably needs a check similar to that provided for the compute jacobian function. it may be wise to add a function that does this check and perform the check after you call `get_labeled_model()`

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,19 +880,125 @@ def _Q_opt(
+        l = len(self.estimated_theta)
+
+        # Assumption: Objective value is sum of squared errors
+        sse = self.objective_value
```
Does this assume the problem is solved? If so, this should be checked somehow?

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,19 +880,125 @@ def _Q_opt(
+                            cov, index=self.estimated_theta.keys(), columns=self.estimated_theta.keys()
+                        )
+                    elif method == "jacobian":
+                        cov = compute_cov(self.exp_list, method, thetavals=self.estimated_theta, solver=solver,
```
This is repeated code, you may collapse these two options into one with an or statement.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,19 +880,125 @@ def _Q_opt(
+        expected value.)
+        '''
+        # get the model
+        model = self.exp_list[0].get_labeled_model()
```
Should this method loop through the experiments? Why is only the `experiment_outputs` of the zero-th model considered?

I could be confusing the methods here but I'm unsure.

djlaky commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -1698,38 +2080,14 @@ def _Q_opt(
-                    cov, index=thetavals.keys(), columns=thetavals.keys()
-                )
+                raise NotImplementedError('Computing the covariance is no longer supported '
+                                          'in the deprecated interface')
```
Why does us changing the new interface change anything that the old interface did? This feels counter-intuitive.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -229,12 +232,458 @@ def _experiment_instance_creation_callback(
+
+class CovMethodLib(Enum):
+    finite_difference = "finite_difference"
+    kaug = "kaug"
```
Upon second throught, I think we should change this to "automatic_differentiation_kaug". The reason is subtle. There is also legacy kaug code that computes the reduced hessian. This change in name also anticipates a future contribution from me for "automatic_differentiation_pynumero". I want to have both versions of AD available for a while until we are confident PyNumero completely agrees with kaug and we fully depreciate kaug (if ever) from the codebase.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -229,12 +232,458 @@ def _experiment_instance_creation_callback(
+    Returns:
+        J: Jacobian matrix
+    """
+    # grab and clone the model
```
@slilonfe5 @djlaky Are these model sanity checks being done elsewhere in ParmEst? If yes, I think we should move them to a private function. This would reduce redundant code in ParmEst.

These sanity checks are probably also already in Pyomo.DoE. I recommend we punt on reducing redundancies between ParmEst and Pyomo.DoE for a future PR.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -229,12 +232,458 @@ def _experiment_instance_creation_callback(
+    Arguments:
+        experiment: Estimator class object, that contains the model for a particular experimental condition
+        thetavals: dictionary containing the estimates of the unknown parameters
+        step: float used to perturb the parameters
```
Specify in the doc string this is a relative perturbation, e.g., step=0.02 is a 2% perturbation.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -279,23 +728,32 @@ def __init__(
+        # check if the experiment has a ``get_labeled_model`` function
         try:
-            outputs = [k.name for k, v in model.experiment_outputs.items()]
+            model = self.exp_list[0].get_labeled_model()
```
This code seems to be repetitive with above. We probably want to modularize it. Maybe we can have a static method _get_labeled_model_helper( ) that takes the `self.exp_list[0]` in this case as an input and returns model. This helper function would contain the value error.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -279,23 +728,32 @@ def __init__(
+        # check that experimental outputs exist
         try:
-            params = [k.name for k, v in model.unknown_parameters.items()]
+            measured_variables = [y_hat.name for y_hat, y in model.experiment_outputs.items()]
```
Same idea here... perhaps use a _get_measured_variables_helper

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -279,23 +728,32 @@ def __init__(
+            measured_variables = [y_hat.name for y_hat, y in model.experiment_outputs.items()]
+        except:
+            raise RuntimeError(
+                "Experiment model does not have suffix " + '"experiment_outputs".'
```
Helper function here?

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -415,9 +713,10 @@ def _create_parmest_model(self, experiment_number):
                     raise RuntimeError(
-                        f"Parmest will not override the existing model component named {n}"
+                        f"Parmest will not override the existing model component named {n}. "
+                        f"Rerun the Estimator object before running theta_est again"
```
I agree with @djlaky. Let's discuss.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -453,13 +914,7 @@ def _instance_creation_callback(self, experiment_number=None, cb_data=None):
-        bootlist=None,
-        calc_cov=False,
-        cov_n=None,
+        self, ThetaVals=None, solver="ef_ipopt", return_values=[], bootlist=None
```
Did you run black? I suspect your change is undoing some black formatting edits.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,19 +1036,169 @@ def _Q_opt(
+        # check if the supplied method is supported
+        try:
+            cov_method = CovMethodLib(method)
+        except:
```
Should this catch a specific type of exception from Enum? As written, this is catching all possible exceptions.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,19 +1036,169 @@ def _Q_opt(
+                            estimated_var=measurement_var,
+                        )
+                    else:
+                        raise NotImplementedError(
```
Perhaps define the NotImplementedError in one spot above? Or at least it's text.

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,19 +1036,169 @@ def _Q_opt(
+                            'supported.'
+                        )
+                else:
+                    raise ValueError(
```
This specific error checking seems repetitive. How can we modularize it?

adowling2 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -882,44 +1447,54 @@ def theta_est(
+        Covariance matrix calculation using all scenarios in the data
+
+        Argument:
+            method: string ``method`` object specified by the user (e.g., kaug)
```
Best practice is to list out all of the supported options here.

adowling2 commented on pyomo/contrib/parmest/tests/test_parmest_cov.py:
```diff
@@ -0,0 +1,2590 @@
+
+# Test class for when the user wants to use the built-in Parmest SSE objective function
+@unittest.skipIf(not ipopt_available, "The 'ipopt' command is not available")
+class TestRooneyBieglerSSE(unittest.TestCase):
```
What is the goal of this file? Does it test both the cov calculations and the new sse objective option? If so, let's consider a more descriptive file name.

slilonfe5 replied:
@adowling2 It's meant to be the test file for the new `SSE_weighted` objective and the cov calculations. I haven't worked on the test file yet. The contents are just duplicates from another test file, awaiting my modifications.

adowling2 commented on pyomo/contrib/parmest/tests/test_parmest_cov.py:
```diff
@@ -0,0 +1,2590 @@
+    @unittest.skipIf(
+        not graphics.imports_available, "parmest.graphics imports are unavailable"
+    )
+    def test_bootstrap(self):
```
Is this retesting bootstrap with SSE?

adowling2 commented on pyomo/contrib/parmest/tests/test_parmest_cov.py:
```diff
@@ -0,0 +1,2590 @@
+        self.assertEqual(bootstrap_theta.shape[0], 3)  # bootstrap for sample 1
+        self.assertEqual(bootstrap_theta[1.0].sum(), 3)  # all true
+
+    def test_diagnostic_mode(self):
```
Is this retesting with SSE?

adowling2 commented on pyomo/contrib/parmest/tests/test_parmest_cov.py:
```diff
@@ -0,0 +1,2590 @@
+
+# Test class for when the user wants to use the built-in Parmest SSE_weighted objective function
+@unittest.skipIf(not ipopt_available, "The 'ipopt' command is not available")
+class TestRooneyBieglerWSSE(unittest.TestCase):
```
@slilonfe5 Do these tests give the same answers for WSSE and SEE? If so, we should talk to the Pyomo team about how to reduce the amount of redundant testing code.

adowling2 reviewed (commented):
@slilonfe5 Next steps:
- @slilonfe5, @djlaky, and @adowling2 discuss today in-person
- @slilonfe5 removes the "draft" label and tags the Pyomo team
- We schedule a call with a subset of the Pyomo team for a preliminary design review. Before the meeting, we should prepare and share an agenda with specific questions for the Pyomo team.
- Keep refining with input

slilonfe5 commented:
@adowling2 @djlaky @jsiirola @blnicho @mrmundt The new capabilities we are introducing in the parmest.py file are ready for review. The test file (test_new_parmest_capabilities.py) is almost ready, I have questions during today's meeting.

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -508,40 +968,32 @@ def _Q_opt(
+
+            # parmest makes the fitted parameters stage 1 variables
+            ind_vars = []
+            for ndname, Var, solval in ef_nonants(ef):
```
`ndname` and `solval` should probably be `nd_name` and `sol_val`. Also I'm not sure what `ef_nonants` is supposed to mean?

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,19 +1040,210 @@ def _Q_opt(
-                return objval, thetavals, cov
-            else:
-                return objval, thetavals
+                return objval, thetavals, var_values
```
Should be `obj_val`

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -719,14 +1337,14 @@ def _Q_at_theta(self, thetavals, initialize_parmest_model=False):
                 if self.diagnostic_mode:
-                    print('      Experiment = ', snum)
-                    print('     First solve with special diagnostics wrapper')
+                    print("      Experiment = ", snum)
+                    print("     First solve with special diagnostics wrapper")
```
It looks like you may have not correctly applied `black.` Reminder that we use `black -S -C .` when running `black` - this stops it from changing `'` to `"`

mrmundt commented:
I did not review the test file at all yet - I wanted to at least give you all initial feedback to look over.

mrmundt commented on pyomo/contrib/parmest/tests/test_new_parmest_capabilities.py:
```diff
@@ -0,0 +1,282 @@
+            "SSE"  # testing the new covariance calculations for the `SSE` objective
+        )
+
+    def check_rooney_biegler_results(self, objval, cov):
+        """
+        Checks if the results are equal to the expected values and agree with the results of Rooney-Biegler
+
+        Argument:
+            objval: the objective value of the annotated Pyomo model
+            cov: covariance matrix of the estimated parameters
```
I would recommend `obj_val` rather than `objval` - it's more accessible on screen readers.

mrmundt commented on pyomo/contrib/parmest/tests/test_new_parmest_capabilities.py:
```diff
@@ -0,0 +1,282 @@
+                    self.create_model()
+                    self.finalize_model()
+                    self.label_model()
+                return self.model
+
+            def create_model(self):
+                m = self.model = rooney_biegler_model()
+
+                return m
+
+            def finalize_model(self):
+                m = self.model
+
+                # fix the input variable
+                m.hour.fix(self.hour)
+
+                return m
+
+            def label_model(self):
+                m = self.model
+
+                # add experiment inputs
+                m.experiment_inputs = pyo.Suffix(direction=pyo.Suffix.LOCAL)
+                m.experiment_inputs.update([(m.hour, self.hour)])
+
+                # add experiment outputs
+                m.experiment_outputs = pyo.Suffix(direction=pyo.Suffix.LOCAL)
+                m.experiment_outputs.update([(m.y, self.y)])
+
+                # add unknown parameters
+                m.unknown_parameters = pyo.Suffix(direction=pyo.Suffix.LOCAL)
+                m.unknown_parameters.update(
+                    (k, pyo.value(k)) for k in [m.asymptote, m.rate_constant]
+                )
+
+                # add measurement error
+                m.measurement_error = pyo.Suffix(direction=pyo.Suffix.LOCAL)
+                m.measurement_error.update([(m.y, None)])
+
+                return m
```
I don't think all of this should live in `setUp`. These will be executed before every single test method, so it's adding overhead that's unnecessary. It'd be better to take them out of the `setUp` method or put them above the `TestModelVariable` class. Something that doesn't make these be re-created with every test.

adowling2 commented:
@mrmundt Thank you for leaving some starter comments. @slilonfe5 and I spoke today about moving the new tests back into the original file and removing the 'new_capabilities' testing file. But your overall comments are helpful!

slilonfe5 commented:
@adowling2 @djlaky @mrmundt @blnicho @jsiirola  
I have implemented the comments on the parmest.py and test_parmest.py files. I moved the test from the previous test_new_parmest_capabilities.py file to the test_parmest.py file. In the test_parmest.py file, the Rooney-Biegler paper example is used to test the 'SSE' and 'SSE_weighted' objectives for all covariance calculation methods ('reduced_hessian', 'finite_difference', and 'automatic_differentiation_kaug'), as well as the two cases of measurement errors (user-supplied and not supplied by the user). To keep the test file short and clean, I didn't consider the reactor design example. Also, I removed the test for the deprecated interface since this will eventually be removed. The above tests passed without issues.

However, I encountered an issue with the test for other functions, such as bootstrap and likelihood ratio. I plan to discuss it today at the Pyomo Dev meeting.

Thanks.

slilonfe5 commented:
@adowling2 @djlaky @mrmundt @blnicho @jsiirola Like I mentioned yesterday, this PR is ready for final review. Also, based on yesterday's meeting, I will create a new PR to fix the bugs in the current parmest.py file and the examples that fail when running the test_examples.py and test_scenariocreator.py files. The new PR will have to go through before this PR goes through.

adowling2 commented:
After creating the new PR, please post a link here.

slilonfe5 commented:
@adowling2 @djlaky @mrmundt @blnicho @jsiirola Here is the link to the new PR on the bug fixes in the `experiment_outputs` and `unknown_parameters` checks in the current parmest.py file, and failing examples in the test_examples.py and test_scenariocreator.py files. https://github.com/Pyomo/pyomo/pull/3635

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -539,51 +1045,113 @@ def _Q_opt(
 
+                self.inv_red_hes = inv_red_hes
+            elif kwargs and all(arg.value in kwargs for arg in UnsupportedArgsLib):
+                deprecation_warning(
+                    "You're using a deprecated call to the `theta_est()` function "
+                    "with the `calc_cov` and `cov_n` arguments. This usage will be "
+                    "removed in a future release. Please update to the new parmest "
+                    "interface using `cov_est()` function for covariance calculation.",
+                    version="6.7.2",
+                )
```
Unless this is an old deprecation warning that you are moving, please change the version to `6.9.3.dev0`.

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,18 +1181,234 @@ def _Q_opt(
+                            tee=self.tee,
+                            estimated_var=measurement_var,
+                        )
+                    else:
+                        raise NotImplementedError(
+                            'Only "finite_difference", "reduced_hessian", and '
+                            '"automatic_differentiation_kaug" methods are supported.'
+                        )
```
This is unreachable code. You already check in lines 1274-1280 whether or not a supported method was supplied, and it raises an error if not. So this is redundant.

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,18 +1181,234 @@ def _Q_opt(
+                            step=step,
+                            tee=self.tee,
+                        )
+                    else:
+                        raise NotImplementedError(
+                            'Only "finite_difference", "reduced_hessian", and '
+                            '"automatic_differentiation_kaug" methods are supported.'
+                        )
```
This is unreachable code. You already check in lines 1274-1280 whether or not a supported method was supplied, and it raises an error if not. So this is redundant.

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -613,18 +1181,234 @@ def _Q_opt(
+                            index=self.estimated_theta.keys(),
+                            columns=self.estimated_theta.keys(),
+                        )
+                    else:
+                        raise NotImplementedError(
+                            'Only "finite_difference", "reduced_hessian", and '
+                            '"automatic_differentiation_kaug" methods are supported.'
+                        )
```
This is unreachable code. You already check in lines 1274-1280 whether or not a supported method was supplied, and it raises an error if not. So this is redundant.

slilonfe5 commented:
@mrmundt Thank you for your feedback. I will work on them to make the code cleaner.

slilonfe5 commented:
@mrmundt @blnicho @jsiirola @adowling2 @djlaky I have implemented Miranda's final review on this PR. I will work on getting this moved to "Done" in the Parmest & Pyomo.DoE Development board. Thanks

lxhowl commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -229,12 +232,496 @@ def _experiment_instance_creation_callback(
+        # calculate the weighted SSE between the prediction and observation of the
+        # measured variables
+        expr = (1 / 2) * sum(
+            ((y - y_hat) / model.measurement_error[y_hat]) ** 2
```
Warning division by zero

blnicho commented:
@slilonfe5 I took a look at the failing doctests and I think you just need to update some code snippets in the parmest documentation to match the API changes in your PR. This is the file and one of the doctests that's failing: https://github.com/Pyomo/pyomo/blame/10d6acc37cca0cf8bc1a5a9734516fc7fd70c989/doc/OnlineDocs/explanation/analysis/parmest/driver.rst#L86

Take a look at all the code snippets in that file and make sure they are using the updated API and won't throw deprecation warnings.

slilonfe5 commented:
@blnicho @mrmundt @jsiirola A kind reminder that I need a final review of the updates I made to the documentation files. Thank you

mrmundt commented on doc/OnlineDocs/explanation/analysis/parmest/covariance.rst:
```diff
@@ -1,16 +1,166 @@
+
+1. Reduced Hessian Method
+
+    When the objective function is "SSE":
```
You mention that SSE is "sum of squared errors" after using it. It's better practice to define the acronym first.

mrmundt commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -26,17 +27,458 @@
 
 
+# Test class for the built-in "SSE" and "SSE_weighted" objective functions
+# validated the results using the Rooney-Biegler paper example
```
Can you include a reference or a link to the paper?

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -517,72 +1057,118 @@ def _Q_opt(
-                        solver_options=self.solver_options,
-                        tee=self.tee,
-                    )
+                # The import error will be raised when we attempt to use
+                # inv_reduced_hessian_barrier below.
+                #
+                # elif not asl_available:
+                #    raise ImportError("parmest requires ASL to calculate the "
+                #                      "covariance matrix with solver 'ipopt'")
```
Unnecessary?

mrmundt commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -517,72 +1057,118 @@ def _Q_opt(
+
+                    solve_result = solver.solve(self.ef_instance, tee=self.tee)
+
+                    # The import error will be raised when we attempt to use
+                    # inv_reduced_hessian_barrier below.
+                    #
+                    # elif not asl_available:
+                    #    raise ImportError("parmest requires ASL to calculate the "
+                    #                      "covariance matrix with solver 'ipopt'")
```
Unnecessary?

slilonfe5 commented:
@mrmundt @blnicho Just a kind reminder that this PR is waiting for Reviewer Approval. Thanks

slilonfe5 commented:
@blnicho Just a kind reminder that I'm waiting for your pending comments (the ones you mentioned during Tuesday's meeting). Thanks

blnicho commented on doc/OnlineDocs/explanation/analysis/parmest/driver.rst:
```diff
@@ -65,40 +67,51 @@ Section.
     # Create an experiment list
     from pyomo.contrib.parmest.examples.rooney_biegler.rooney_biegler import RooneyBieglerExperiment
+
+    class NewRooneyBieglerExperiment(RooneyBieglerExperiment):
+        def label_model(self):
+            m = self.model
+
+            # create the experiment outputs
+            m.experiment_outputs = pyo.Suffix(direction=pyo.Suffix.LOCAL)
+            m.experiment_outputs.update(
+                [(m.response_function[self.data['hour']], self.data['y'])]
+            )
+
+            # create the unknown parameters
+            m.unknown_parameters = pyo.Suffix(direction=pyo.Suffix.LOCAL)
+            m.unknown_parameters.update((k, pyo.value(k)) for k in [m.asymptote, m.rate_constant])
+
+            # create the measurement error
+            m.measurement_error = pyo.Suffix(direction = pyo.Suffix.LOCAL)
+            m.measurement_error.update([(m.response_function[self.data['hour']], None)])
```
Same comment about adding the `measurement_error` suffix to the existing `RooneyBieglerExperiment` class and then just importing it here.

slilonfe5 replied:
@blnicho Similarly, if we are using the new API with built-in objectives, apart from adding `measurement_error` to the current class, the `experiment_outputs` Suffix needs to also be updated. Currently, the class `experiment_outputs` Suffix is updated with both input (`m.hour`) and output (`m.y`) data, without capturing the model predicted outputs (`m.response_function`).

blnicho reviewed (changes requested):
@slilonfe5 here are some initial comments/suggestions. I'm still going through a couple of the files but I have some more substantial questions that I'll message you separately about.

slilonfe5 commented on doc/OnlineDocs/explanation/analysis/parmest/covariance.rst:
```diff
@@ -1,16 +1,176 @@
+    However, this method uses the model optimality (KKT) condition to compute the
+    Jacobian matrix, :math:`\mathbf{G}_{\text{kaug},\, i}`, for experiment :math:`i`.
+
+In parmest, the covariance matrix can be calculated after defining the
+:class:`~pyomo.contrib.parmest.parmest.Estimator` object and estimating the unknown
+parameters using :class:`~pyomo.contrib.parmest.parmest.Estimator.theta_est`. To
+estimate the covariance matrix, call
+:class:`~pyomo.contrib.parmest.parmest.Estimator.cov_est` and pass it the number
+of data points, e.g.,
```
@blnicho Yes. I had this on my To-do list. Thank you for the comment

slilonfe5 commented on doc/OnlineDocs/explanation/analysis/parmest/covariance.rst:
```diff
@@ -1,16 +1,176 @@
+            self.y = y
+            self.hour = hour
+            self.model = None
+
+        def get_labeled_model(self):
+            self.create_model()
+            self.finalize_model()
+            self.label_model()
+
+            return self.model
+
+        def create_model(self):
+            m = self.model = rooney_biegler_model()
+
+            return m
+
+        def finalize_model(self):
+            m = self.model
+
+            # fix the input variable
+            m.hour.fix(self.hour)
+
+            return m
+
+        def label_model(self):
+            m = self.model
+
+            # add experiment outputs
+            m.experiment_outputs = pyo.Suffix(direction=pyo.Suffix.LOCAL)
+            m.experiment_outputs.update([(m.y, self.y)])
+
+            # add unknown parameters
+            m.unknown_parameters = pyo.Suffix(direction=pyo.Suffix.LOCAL)
+            m.unknown_parameters.update(
+                (k, pyo.value(k)) for k in [m.asymptote, m.rate_constant]
+            )
+
+            # create the measurement error
+            m.measurement_error = pyo.Suffix(direction = pyo.Suffix.LOCAL)
+            m.measurement_error.update([(m.y, None)])
```
The model of the existing experiment class doesn't define the experiment output (i.e., y) as a variable. Instead, it uses a Pyomo Expression, `m.response_function` to calculate what the experiment output (i.e., y) should be. The `cov_est` function requires that the experiment outputs are defined as variables. In addition to a missing `measurement_error` Suffix, the current experiment class also misses an `experiment_outputs` Suffix. These issues can be addressed in a new PR after this one gets merged.

blnicho commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -26,17 +27,419 @@
+
+        self.exp_list = exp_list
+
+        if self.objective_function == "incorrect_obj":
+            with pytest.raises(
+                ValueError,
+                match=r"Invalid objective function: 'incorrect_obj'\. "
+                r"Choose from: \['SSE', 'SSE_weighted'\]\.",
+            ):
+                self.pest = parmest.Estimator(
+                    self.exp_list, obj_function=self.objective_function, tee=True
+                )
```
The basic parmest tests should already be testing for invalid input values to the `Estimator` constructor. I think you should rework this option/test to check the behavior when a user supplies a custom objective function. In that case you shouldn't see an error until you try to run `cov_est`.

blnicho commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -26,17 +27,419 @@
+            ("unsupported_method"),
+        ]
+    )
+    def test_parmest_covariance(self, cov_method):
```
The if-tree in this test seems overly complicated. You can simplify it by checking for the special cases that you expect to raise an error first and then the rest of the cases all follow the same steps.

slilonfe5 replied:
I don't understand what you suggest here. We are testing the covariance results of all the methods

blnicho commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -254,16 +656,16 @@ def test_theta_est_cov(self):
+            cov["asymptote"]["asymptote"], 6.155892, places=2
         )  # 6.22864 from paper
         self.assertAlmostEqual(
-            cov["asymptote"]["rate_constant"], -0.4395341, places=2
+            cov["asymptote"]["rate_constant"], -0.425232, places=2
         )  # -0.4322 from paper
         self.assertAlmostEqual(
-            cov["rate_constant"]["asymptote"], -0.4395341, places=2
+            cov["rate_constant"]["asymptote"], -0.425232, places=2
         )  # -0.4322 from paper
         self.assertAlmostEqual(
-            cov["rate_constant"]["rate_constant"], 0.04124, places=2
+            cov["rate_constant"]["rate_constant"], 0.040571, places=2
```
Why did these values change?

slilonfe5 replied:
When I updated the `RooneyBieglerExperiment` class to use `Var` and `Constraint`, the covariance results minimally changed.

blnicho reviewed (commented):
@slilonfe5 here are a few more review comments. I'm still going through `parmest.py`

slilonfe5 commented on doc/OnlineDocs/explanation/analysis/parmest/covariance.rst:
```diff
@@ -1,16 +1,114 @@
+
+In parmest, the covariance matrix can be calculated after defining the
+:class:`~pyomo.contrib.parmest.parmest.Estimator` object and estimating the unknown
+parameters using :class:`~pyomo.contrib.parmest.parmest.Estimator.theta_est`. To
+estimate the covariance matrix, call
+:class:`~pyomo.contrib.parmest.parmest.Estimator.cov_est` and pass it the number
+of data points, e.g.,
```
Yes. It is required. The user may not want to use all the data for the calculations.

slilonfe5 commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -26,17 +27,419 @@
+                        0.041242,
+                        places=2,
+                    )  # 0.04124 from paper
+                else:
+                    self.assertAlmostEqual(
+                        cov.iloc[asymptote_index, asymptote_index], 36.935351, places=2
+                    )  # 6.22864 from paper
+                    self.assertAlmostEqual(
+                        cov.iloc[asymptote_index, rate_constant_index],
+                        -2.551392,
+                        places=2,
+                    )  # -0.4322 from paper
+                    self.assertAlmostEqual(
+                        cov.iloc[rate_constant_index, asymptote_index],
+                        -2.551392,
+                        places=2,
+                    )  # -0.4322 from paper
+                    self.assertAlmostEqual(
+                        cov.iloc[rate_constant_index, rate_constant_index],
+                        0.243428,
+                        places=2,
+                    )  # 0.04124 from paper
```
We are using the true `sse` to compute the covariance matrix. Previously, we calculated the covariance by assuming that the `obj_val` is the `sse`, which is not true.

blnicho replied:
So the three calculation methods aren't approximating the same sse and covariance?

slilonfe5 commented:
@blnicho @mrmundt @jsiirola @djlaky @adowling2 I have implemented Bethany's final review.

slilonfe5 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -229,12 +232,548 @@ def _experiment_instance_creation_callback(
+    # grab the model
+    model = _get_labeled_model_helper(experiment)
+
+    # check if the model has all the required suffixes
+    _check_model_labels_helper(model, logging_level)
```
@blnicho You can only reach the `_compute_jacobian` function when you call `SSE` or `SSE_weighted`. The function is used for covariance matrix calculation, which is only supported with the built-in objective functions. The `_check_model_labels_helper` is really needed. It checks and ensures that the model is labeled properly with `experiment_outputs` and `unknown_parameters`.

slilonfe5 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -229,12 +232,548 @@ def _experiment_instance_creation_callback(
+            f"model is well-posed. The original error was {e}."
+        )
+
+    # add zero (dummy/placeholder) objective function
+    if not hasattr(model, "objective"):
+        model.objective = pyo.Objective(expr=0, sense=pyo.minimize)
+
+    solver.solve(model, tee=tee)
```
@blnicho I've spoken with Dan about this. This is not needed

slilonfe5 commented:
@blnicho @mrmundt Just a kind reminder that we need this PR to be merged ahead of the PSE Summit next week.

mrmundt commented on pyomo/contrib/parmest/tests/test_parmest.py:
```diff
@@ -488,6 +773,21 @@ def create_model(self):
+                    [(m.hour, self.data["hour"]), (m.y, self.data["y"])]
+                )
+
+                m.unknown_parameters = pyo.Suffix(direction=pyo.Suffix.LOCAL)
+                m.unknown_parameters = pyo.Suffix(direction=pyo.Suffix.LOCAL)
```
Am I crazy, or are these the same line repeated twice? Is this intentional?

slilonfe5 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -229,12 +233,569 @@ def _experiment_instance_creation_callback(
+        raise RuntimeError(f"Failed to clone labeled model: {exc}")
+
+
+def _check_solver_termination_result(model, solver_result):
+    """
+    Checks if the Pyomo model solves to optimality
+    """
+    if solver_result.solver.termination_condition == TerminationCondition.optimal:
+        model.solutions.load_from(solver_result)
+    else:
+        logger.error(
+            "Solution is not optimal. Termination condition: %s, Status: %s",
+            solver_result.solver.termination_condition,
+            solver_result.solver.status,
+        )
+        raise RuntimeError(
+            f"Model from experiment did not solve appropriately. "
+            f"Make sure the model is well-posed."
+        )
```
@mrmundt Thank you

mrmundt reviewed (approved):
There are some refactoring opportunities, but from a software engineering perspective, this is good enough to merge at this point.

slilonfe5 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -425,8 +978,12 @@ def _create_parmest_model(self, experiment_number):
+                    second_stage_rule = SSE
+                else:
+                    second_stage_rule = SSE_weighted
+                self.covariance_objective = second_stage_rule
```
@jsiirola Yes. This is intentional. The covariance matrix calculation is only valid for the Enum objective functions

slilonfe5 commented on pyomo/contrib/parmest/parmest.py:
```diff
@@ -229,12 +232,531 @@ def _experiment_instance_creation_callback(
+        )
+
+    try:
+        return get_model().clone()
```
@jsiirola I have implemented all the comments except this one. I agree with you, the `compute_jacobian` and other covariance matrix functions can be made not to clone the model. I don't have enough time (I don't know if there is enough time) to do this. I can note it as something to work on after the release.

mrmundt commented:
> Alternatively, the `if` and `elif` statements can be put inside the first `if` statement that checks if `calc_cov is not NOTSET`

We resolved it a bit differently, but it should be fine now.

slilonfe5 commented:
@mrmundt I can attempt to solve all the testing issues with the structure John recommended, and you can review it afterwards

mrmundt commented:
> @mrmundt I can attempt to solve the testing issues with the structure John recommended, and you can review it afterwards

Should be fine now. I had to update some docs.

mrmundt commented:
> > > > > @mrmundt I can attempt to solve the testing issues with the structure John recommended, and you can review it afterwards
> > > > 
> > > > 
> > > > Should be fine now. I had to update some docs.
> > > 
> > > 
> > > With the current updates to the doc files, we shouldn't have a deprecation warning because we are using the string `obj_function` input and also not passing a `calc_cov` argument to `theta_est()`
> > 
> > 
> > Are you able to address the comments today, ASAP? We really do need to cut the release.
> 
> Yes. I can do that. Thanks

Okay, so, the error that you have now is the same one we were hitting, which is why we added the second deprecation warning. What's your preferred path forward?

slilonfe5 commented:
@mrmundt, we may need to re-run the one test that failed on the Python packages installation

slilonfe5 commented:
> > @mrmundt, we may need to re-run the one test that failed on the Python packages installation
> 
> Yup, will do. I have to wait for all the other tests to finish first.

Thank you for all your help. Hopefully, we get to connect during the PSE Summit next week

mrmundt commented:
> > > @mrmundt, we may need to re-run the one test that failed on the Python packages installation
> > 
> > 
> > Yup, will do. I have to wait for all the other tests to finish first.
> 
> Thank you for all your help. Hopefully, we get to connect during the PSE Summit next week

Oooop. Hate to disappoint, but I won't be there! John will be, though. (If you make whatever discussion a Zoom meeting, I'm happy to call in virtually, FWIW.)
